src/kernel/fs.zig

```diff
@@ -0,0 +1,661 @@
+    ///
+    /// Error: std.mem.Allocator.Error
+    ///     std.mem.Allocator.Error.OutOfMemory - There wasn't enough memory to fulfill the request
+    ///
+    const Read = fn (self: *const Self, node: *const FileNode, len: usize) std.mem.Allocator.Error![]u8;
+    ///
+    /// Write to an open file
+    ///
+    /// Arguments:
+    ///     IN self: *const FileSystem - The filesystem in question being operated on
+    ///     IN node: *const FileNode - The file being read from
+    ///     IN bytes: []u8 - The bytes to write to the file
+    ///
+    /// Error: std.mem.Allocator.Error
+    ///     std.mem.Allocator.Error.OutOfMemory - There wasn't enough memory to fulfill the request
+    ///
+    const Write = fn (self: *const Self, node: *const FileNode, bytes: []const u8) std.mem.Allocator.Error!void;
+    ///
+    /// Open a file/dir within the filesystem. The result can then be used for write, read or close operations
+    ///
+    /// Arguments:
+    ///     IN self: *const FileSystem - The filesystem in question being operated on
+    ///     IN node: *const DirNode - The directory under which to open the file/dir from
+    ///     IN name: []const u8 - The name of the file to open
+    ///     IN flags: OpenFlags - The flags to consult when opening the file
+    ///
+    /// Return: *const Node
+    ///     The node representing the file/dir opened
+    ///
+    /// Error: std.mem.Allocator.Error || Error
+    ///     std.mem.Allocator.Error.OutOfMemory - There wasn't enough memory to fulfill the request
+    ///     Error.NoSuchFileOrDir - The file/dir by that name doesn't exist and the flags didn't specify to create it
+    ///
+    const Open = fn (self: *const Self, node: *const DirNode, name: []const u8, flags: OpenFlags) (std.mem.Allocator.Error || Error)!*Node;
+
+    ///
+    /// Get the node representing the root of the filesystem. Used when mounting to bind the mount point to the root of the mounted fs
+    ///
+    /// Arguments:
+    ///     IN self: *const Self - The filesystem to get the root node for
+    ///
+    /// Return: *const DirNode
+    ///     The root directory node
+    ///
+    const GetRootNode = fn (self: *const Self) *const DirNode;
+
+    /// The close function
+    close: Close,
+    /// The read function
+    read: Read,
+    /// The write function
+    write: Write,
+    /// The open function
+    open: Open,
+    getRootNode: GetRootNode,
+
+    /// Points to a usize field within the underlying filesystem so that the close, read, write and open functions can access its low-level implementation using @fieldParentPtr. For example, this could point to a usize field within a FAT32 filesystem data structure, which stores all the data and state that is needed in order to interact with a physical disk
+    /// The value of instance is reserved for future use and so should be left as 0
+    instance: *usize,
```

Comment:
I don't think you need a instance pointer, as `@fieldParentPtr` can use the struct point itself. e.g. in your `TestFS` when you pass a `fs: *const FileSystem`, you can just do: `@fieldParentPtr(TestFs, "fs", fs)`, instead of:  `@fieldParentPtr(TestFs, "instance", fs.instance)`. But not sure you can do this with your implementation as there will need to change pointer types and idk if it will be more of a mess. I tried doing this, but will need to remove the `*const` pointers and idk if this should be kept as const pointers.

Reply:
I think it is better with the const pointers and less mess, personally :)

Reply:
I was thinking that :) So keep as is

Reply:
Noice